powdr/src/lib.rs

```diff
@@ -12,3 +12,179 @@ pub use powdr_pipeline::Pipeline;
 pub use powdr_number::GoldilocksField;
+
+use riscv::CompilerOptions;
+
+use std::fs::File;
+use std::path::Path;
+use std::path::PathBuf;
+use std::time::Instant;
+
+pub struct Session {
+    pipeline: Pipeline<GoldilocksField>,
+    out_path: String,
+}
+
+const DEFAULT_PKEY: &str = "pkey.bin";
+const DEFAULT_VKEY: &str = "vkey.bin";
+
+impl Session {
+    pub fn new(guest_path: &str, out_path: &str) -> Self {
+        Session {
+            pipeline: pipeline_from_guest(guest_path, Path::new(out_path)),
+            out_path: out_path.into(),
+        }
+        .with_backend(powdr_backend::BackendType::Plonky3)
+    }
+
+    pub fn into_pipeline(self) -> Pipeline<GoldilocksField> {
+        self.pipeline
+    }
+
+    pub fn pipeline(&self) -> &Pipeline<GoldilocksField> {
+        &self.pipeline
+    }
+
+    pub fn with_backend(self, backend: backend::BackendType) -> Self {
+        Session {
+            pipeline: self.pipeline.with_backend(backend, None),
+            ..self
+        }
+    }
+
+    pub fn write<S: serde::Serialize>(self, channel: u32, data: &S) -> Self {
+        Session {
+            pipeline: self.pipeline.add_data(channel, data),
+            ..self
+        }
+    }
+
+    pub fn run(&mut self) {
+        run(&mut self.pipeline);
+    }
+
+    pub fn prove(&mut self) {
+        match self.pipeline.read_constants_mut(Path::new(&self.out_path)) {
+            Ok(_) => (),
+            Err(_) => {
+                self.pipeline.compute_fixed_cols().unwrap();
+            }
+        }
+
+        let pkey = Path::new(&self.out_path).join(DEFAULT_PKEY);
+        let vkey = Path::new(&self.out_path).join(DEFAULT_VKEY);
+
+        if pkey.exists() && vkey.exists() {
+            self.pipeline.set_pkey_file(pkey.clone());
+            self.pipeline.set_vkey_file(vkey.clone());
+            self.pipeline.setup_backend().unwrap();
+        } else {
+            self.pipeline.setup_backend().unwrap();
+            self.export_setup();
+            self.pipeline.set_pkey_file(pkey.clone());
+            self.pipeline.set_vkey_file(vkey.clone());
+        }
+
+        prove(&mut self.pipeline);
+    }
+
+    pub fn export_setup(&mut self) {
+        let mut path = PathBuf::from(self.out_path.clone());
+        path.push(DEFAULT_PKEY);
+        let file = File::create(path).unwrap();
+
+        self.pipeline.export_proving_key(file).unwrap();
+
+        let mut path = PathBuf::from(self.out_path.clone());
+        path.push(DEFAULT_VKEY);
+        let file = File::create(path).unwrap();
+
+        self.pipeline.export_verification_key(file).unwrap();
+    }
+}
+
+pub fn build_guest(guest_path: &str, out_path: &Path) -> (PathBuf, String) {
+    riscv::compile_rust(
+        guest_path,
+        CompilerOptions::new_gl()
+            .with_poseidon()
+            .with_continuations(),
+        out_path,
+        true,
+        None,
+    )
+    .ok_or_else(|| vec!["could not compile rust".to_string()])
+    .unwrap()
```

Comment:
bit rude, could this be due to a user error?

Reply:
I guess it could, if the Rust code itself doesn't compile. But in that case the user will also see the output from the Rust compiler which is pretty clear.